Initial setup of QA pre-commit tools [cf-units]

ukmo-ccbunney opened the pull request:
## 🚀 Pull Request

### Description
Added following pre-commit hooks:
  - Extra "default" hooks from pre-commit (trailing whitespace, etc)
  - MyPy
  - codespell
  - validate_pyproject
  - numpydoc

`sp-repo-review` was already setup for cf-units, but have updated the ignore list as various tests now pass.

`ruff` was enabled, but with only selected tests. Have enabled "all"  tests.

For all QA tools above, the results of any "auto-fixes" have been included in this repo, mainly fixes for:
  - trailing whitespace
  - missing newline at end of file
  - docstring reformatting
  - some code style linting from ruff

All other failing tests have been added to ignore lists. See separate issues:
   - #520 
   - #521
   - #526 
   - #462

ukmo-ccbunney committed:
Added extra pre-commit default hooks

ukmo-ccbunney committed:
Added codespell pre-commit hook

ukmo-ccbunney committed:
Added mypy and configuration

ukmo-ccbunney committed:
Added mypy exrts dependency

ukmo-ccbunney committed:
Added ignores for mypy (including one in sp-repo-review)

ukmo-ccbunney committed:
Added validate-pyproject

ukmo-ccbunney committed:
Added numpydoc precommit hook and ignored errors in pyproject.toml

ukmo-ccbunney committed:
Updated ignore_list of sp-repo-review for errors that have been
addressed in this branch

ukmo-ccbunney committed:
Fixed sp-repo-review "PC901" (Custom pre-commit CI message)

ukmo-ccbunney committed:
Added extra pytest config, as recommended by sp-repo-review.

ukmo-ccbunney committed:
Added `filterwarnings=error` option for pytest and an exception for thie
existing known Deprecation warnings

ukmo-ccbunney committed:
- Enabled all Ruff checks.
  - Required lots of additions to the Ruff ignore list
  - Extra additions to mypy ignore list
- Autofixes from Ruff linter
- Fixes from codespell

ukmo-ccbunney committed:
Tidy up pyproject.toml.
  - Removed some duplicate ruff ignores
  - Reorded some subsections

ukmo-ccbunney committed:
Updated Ruff config:
  - Added: `force-sort-within-sections = true`
  - Auto updated files resultig from above
  - file specific ignores

ukmo-ccbunney committed:
Auto-corrections for
  - Trailing witespace
  - No neline at end of file
  - Some Ruff linting

pp-mo commented on cf_units/_udunits2_parser/_antlr4_runtime/Recognizer.py:
```diff
@@ -7,7 +7,7 @@
 from .Token import Token
 
-# need forward delcaration
+# need forward declaration
```
It's not appropriate to modify **anything** inside `cf_units._udunits_parser._antlr4_runtime` ,
since this code is all copied directly from another project.

Effectively this is generated code, derived from the action of [compile.py](https://github.com/SciTools/cf-units/blob/v3.3.0/cf_units/_udunits2_parser/compile.py).
In a more ideal world, this code wouldn't exist in this repo, and we'd have it as a importable dependency instead.
It's only included in the project repo for convenience, specifically to avoid an awkward dependency.
See [./_udunits2_parser/README.md#generating-the-parser-from-the-grammar](https://github.com/SciTools/cf-units/blob/v3.3.0/cf_units/_udunits2_parser/README.md#generating-the-parser-from-the-grammar) for the lowdown.

So, we need to exclude this from the relevant checks.
I _think_ you can just add a top-level [exclude](https://pre-commit.com/#top_level-exclude) control to `.pre-commit-config.yaml`.

ukmo-ccbunney replied:
Rolled back changes to files in `_antlr4_runtime` directory and added top level exclusion in `pre-commit-config.yaml`; see 89afe5a

pp-mo commented on .pre-commit-config.yaml:
```diff
@@ -2,6 +2,10 @@
 # See https://pre-commit.com/hooks.html for more hooks
 # See https://github.com/scientific-python/cookie#sp-repo-review for repo-review
 
+ci:
```
For consistency with the "repo procedures" templates in the project, 
I think this should also have a reference line `# See https://pre-commit.ci/#configuration` in the initial comment lines section above.

ukmo-ccbunney replied:
Good spot; fixed in 2749787

ukmo-ccbunney committed:
- Update Ruff version.
- Add trailing / to mypy exclude paths

pp-mo commented on .pre-commit-config.yaml:
```diff
@@ -16,8 +20,21 @@ repos:
     -   id: check-merge-conflict
         # Check for debugger imports and py37+ `breakpoint()` calls in Python source.
     -   id: debug-statements
+        # Check TOML file syntax.
```
It could be useful to feed these explainers back to the template, with a PR to [the 'template' file in to SciTools/.github](https://github.com/SciTools/.github/blob/6cd87da89ae606ccb4ce0a4c95421c012cb29108/templates/.pre-commit-config.yaml)

ukmo-ccbunney replied:
I don't follow you here - the comments already exist in the template file.

pp-mo replied:
Oh I hadn't seen that !

ukmo-ccbunney committed:
Try different approach to MyPy ignore files in .pre-commit config

ukmo-ccbunney committed:
Update pyproject.toml

Added extra ignores in mypy for _udunits generated code

ukmo-ccbunney committed:
Update pyproject.toml

Re-added PY005 test to sp-repo-review ignore list.

pp-mo commented on pyproject.toml:
```diff
@@ -69,6 +69,10 @@ Discussions = "https://github.com/SciTools/cf-units/discussions"
 Documentation = "https://cf-units.readthedocs.io"
 
+[tool.codespell]
+#ignore-words-list = "Wirth"
```
What is this bit, can we either remove or explain ?

ukmo-ccbunney replied:
Redundant exception - now removed; 843e6c2

pp-mo commented on pyproject.toml:
```diff
@@ -109,33 +119,121 @@ local_scheme = "dirty-tag"
 
 [tool.ruff.lint]
+ignore = [
+    # NOTE: Non-permanent exclusions should be added to the ".ruff.toml" file.
```
Sorry to hit you with this, but I think the sprint team has now definitely decided **_not_** to do this. 
So, should remove this : the approach at [line 145](https://github.com/SciTools/cf-units/blob/7e9a9bb34bbcc3219391682cf12e7cebaf16dadd/pyproject.toml#L145) is now the way !

ukmo-ccbunney replied:
Yep! Old comment. I've got nothing in .ruff.toml anyhow. Removed: 5a955ff

pp-mo commented on pyproject.toml:
```diff
@@ -145,22 +243,110 @@ convention = "numpy"
+    # TODO: exceptions that still need investigating are below.
+    # Might be fixable, or might become permanent (above):
+    "PY005",  # Has tests folder
+    "PP003",  # Does not list wheel as a build-dep
```
FWIW I think it might be useful to keep these in alphabetical order.
Look like that is true of the existing, anyway

ukmo-ccbunney replied:
They were in order that repo-review reports them but it makes sense to alphabetise them for consistency; see 01fe1b7

pp-mo commented on pyproject.toml:
```diff
@@ -109,33 +119,121 @@ local_scheme = "dirty-tag"
-]
-ignore = ["B904", "F403", "F405"]
+    "ALL",
+    "D212",  # Multi-line docstring summary should start at the first line
```
The D212 entry appears to be redundant ?

ukmo-ccbunney replied:
This was explicitly set in the standard scitools linter config (https://github.com/orgs/SciTools/projects/43/views/3?pane=issue&itemId=85886972), so I've left it in?

pp-mo replied:
Yes, but I'm pretty sure it's wrong all the same.

To my frustration, they don't seem to actually document the "ALL" magic 😞 
But IIUC, the list of 'select' options is additive, so including "ALL" means anything else is redundant by definition.

(NB we have an ongoing offline teams convo about this)

pp-mo replied:
> they don't seem to actually document the "ALL" magic

Oh my mistake, it is there in the [command-line help](https://docs.astral.sh/ruff/configuration/#full-command-line-interface)

ukmo-ccbunney replied:
D212 removed from Ruff `select`; bb551e1
This resulted in a "unused NOQA" autofix on `__init__.py` which is also included in the commit.

pp-mo commented on pyproject.toml:
```diff
@@ -109,33 +119,121 @@ local_scheme = "dirty-tag"
-    "UP",
-]
-ignore = ["B904", "F403", "F405"]
+    "ALL",
```
It seems odd to me to have "exclude" followed by "select", since 'exclude' effectively qulaifies the 'select' choices.  
( but I've now raised this offline in our Teams sprint chat )

ukmo-ccbunney replied:
Will raise a separate PR for this once we have a decision?

pp-mo reviewed (changes requested):
Ok but a few nit-picky suggestions.

You should also create some PRs to track unfinished business.
Suggest one each for :
  * re-order pre-commit entries based on [updated tempate file](https://github.com/SciTools/.github/pull/29)  : (can do this one right now)
  * mypy
  * ruff
  * numpydocs-validation

ukmo-ccbunney committed:
Roll back changes to _antlr4_runtime and add .pre-commit exclusion for
this directory

ukmo-ccbunney committed:
Added reference comment to .pre-commit-config.yaml

ukmo-ccbunney committed:
Removed redundant setting

ukmo-ccbunney committed:
Removed comment

ukmo-ccbunney committed:
Alphabetise repo-review ignore list

ukmo-ccbunney committed:
Ahphabetised pre-commit hooks, as per template

ukmo-ccbunney committed:
Merge branch 'main' into qa_reresh_pre-commit-hooks

ukmo-ccbunney committed:
Added explanation comments to Ruff ignores

ukmo-ccbunney committed:
Added reference to common SciTools linting rules Wiki

pp-mo reviewed (changes requested):
I think removing "D212" from the ruff.lint `select` is worth doing,
not least because the comment is confusing + I think wrong!

otherwise all looks great now + good to go !

ukmo-ccbunney committed:
Removed D212 from Ruff select

ukmo-ccbunney committed:
Reinstate D212 Ruff rule.
This also updated one docstring in cf_units/__init__.py (correctly) due
to the removal of a redundant `noqa: E501` marker in a previous commit
